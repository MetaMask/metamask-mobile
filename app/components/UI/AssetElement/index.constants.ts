export const FIAT_BALANCE_TEST_ID = 'fiat-balance-test-id';
export const MAIN_BALANCE_TEST_ID = 'main-balance-test-id';
