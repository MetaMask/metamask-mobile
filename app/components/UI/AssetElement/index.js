import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { TouchableOpacity, StyleSheet, View } from 'react-native';
import Icon from 'react-native-vector-icons/Ionicons';
import { ThemeContext, mockTheme } from '../../../util/theme';

const createStyles = (colors) =>
  StyleSheet.create({
    itemWrapper: {
      flex: 1,
      flexDirection: 'row',
      paddingHorizontal: 15,
      paddingVertical: 10,
      borderBottomWidth: StyleSheet.hairlineWidth,
      borderBottomColor: colors.border.muted,
    },
    arrow: {
      flex: 1,
      alignSelf: 'flex-end',
    },
    arrowIcon: {
      marginTop: 16,
    },
  });

/**
 * Customizable view to render assets in lists
 */
export default class AssetElement extends PureComponent {
  static propTypes = {
    /**
     * Content to display in the list element
     */
    children: PropTypes.node,
    /**
     * Object being rendered
     */
    asset: PropTypes.object,
    /**
     * Callback triggered on long press
     */
    onPress: PropTypes.func,
    /**
     * Callback triggered on long press
     */
    onLongPress: PropTypes.func,
  };

  handleOnPress = () => {
    const { onPress, asset } = this.props;
    onPress && onPress(asset);
  };

  handleOnLongPress = () => {
    const { onLongPress, asset } = this.props;
    onLongPress && onLongPress(asset);
  };

  render = () => {
    const { children } = this.props;
    const colors = this.context.colors || mockTheme.colors;
    const styles = createStyles(colors);

    return (
      <TouchableOpacity
        onPress={this.handleOnPress}
        onLongPress={this.handleOnLongPress}
        style={styles.itemWrapper}
      >
        {children}
        <View styles={styles.arrow}>
          <Icon
            name="ios-arrow-forward"
            size={24}
            color={colors.icon.alternative}
            style={styles.arrowIcon}
          />
        </View>
      </TouchableOpacity>
    );
  };
}

AssetElement.contextType = ThemeContext;
