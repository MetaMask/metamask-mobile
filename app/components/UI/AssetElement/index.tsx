/* eslint-disable react/prop-types */
import React from 'react';
import { TouchableOpacity, StyleSheet, Platform } from 'react-native';
import Text, {
  TextVariant,
} from '../../../component-library/components/Texts/Text';
import SkeletonText from '../FiatOnRampAggregator/components/SkeletonText';
import { TokenI } from '../Tokens/types';
import generateTestId from '../../../../wdio/utils/generateTestId';
import { getAssetTestId } from '../../../../wdio/screen-objects/testIDs/Screens/WalletView.testIds';
import {
  TOKEN_BALANCE_LOADING,
  TOKEN_RATE_UNDEFINED,
} from '../Tokens/constants';
interface AssetElementProps {
  children?: React.ReactNode;
  asset: TokenI;
  onPress?: (asset: TokenI) => void;
  onLongPress?: ((asset: TokenI) => void) | null;
  balance?: string;
}

const createStyles = () =>
  StyleSheet.create({
    itemWrapper: {
      flex: 1,
      flexDirection: 'row',
      paddingHorizontal: 15,
      paddingVertical: 10,
      alignItems: 'flex-start',
    },
    arrow: {
      flex: 1,
      alignSelf: 'flex-end',
    },
    arrowIcon: {
      marginTop: 16,
    },
    skeleton: {
      width: 50,
    },
  });

/**
 * Customizable view to render assets in lists
 */
const AssetElement: React.FC<AssetElementProps> = ({
  children,
  balance,
  asset,
  onPress,
  onLongPress,
}) => {
  const styles = createStyles();

  const handleOnPress = () => {
    onPress?.(asset);
  };

  const handleOnLongPress = () => {
    onLongPress?.(asset);
  };

  return (
    <TouchableOpacity
      onPress={handleOnPress}
      onLongPress={handleOnLongPress}
      style={styles.itemWrapper}
      {...generateTestId(Platform, getAssetTestId(asset.symbol))}
    >
      {children}

      {balance && (
        <Text
          variant={
            asset?.balanceError || asset.balanceFiat === TOKEN_RATE_UNDEFINED
              ? TextVariant.BodySM
              : TextVariant.BodyLGMedium
          }
        >
          {balance === TOKEN_BALANCE_LOADING ? (
            <SkeletonText thin style={styles.skeleton} />
          ) : (
            balance
          )}
        </Text>
      )}
    </TouchableOpacity>
  );
};
export default AssetElement;
