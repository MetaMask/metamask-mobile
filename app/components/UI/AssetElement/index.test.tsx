import React from 'react';
import { shallow } from 'enzyme';
import AssetElement from './';

describe('AssetElement', () => {
  it('should render correctly', () => {
    const wrapper = shallow(
      <AssetElement
        asset={{
          name: 'name',
          symbol: 'symbol',
          description: 'description',
          address: '0x123',
          totalSupply: 1,
        }}
      />,
    );
    expect(wrapper).toMatchSnapshot();
  });
});
