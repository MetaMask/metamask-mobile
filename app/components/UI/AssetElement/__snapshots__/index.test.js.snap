// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AssetElement should render correctly 1`] = `
<TouchableOpacity
  activeOpacity={0.2}
  onLongPress={[Function]}
  onPress={[Function]}
  style={
    Object {
      "borderBottomColor": "#CCCCCC",
      "borderBottomWidth": 0.5,
      "flex": 1,
      "flexDirection": "row",
      "paddingHorizontal": 15,
      "paddingVertical": 10,
    }
  }
>
  <View
    styles={
      Object {
        "alignSelf": "flex-end",
        "flex": 1,
      }
    }
  >
    <Icon
      allowFontScaling={false}
      color="#AAAAAA"
      name="ios-arrow-forward"
      size={24}
      style={
        Object {
          "marginTop": 16,
        }
      }
    />
  </View>
</TouchableOpacity>
`;
